include/targets/cuda/mma_tensor_op/gmem_loader.cuh

```diff
@@ -0,0 +1,792 @@
+                                            smem_accessor_t &smem_imag)
+      {
+        // for each iteration, each warp loads a tile
+        int thread_id = (threadIdx.z * blockDim.y + threadIdx.y) * blockDim.x + threadIdx.x;
+        int warp_id = thread_id / 32;
+        int lane_id = thread_id % 32;
+        constexpr bool x = transpose == dagger;
+        constexpr int tmp_ld = x ? bN + get_tmp_pad() : bM + get_tmp_pad();
+        tmp2s_smem_t<T, load_t, x, tmp_ld, 1> pattern; // 1 is dummy here
+        constexpr int w_m = pattern.get_wm();
+        constexpr int w_k = pattern.get_wn();
+        static_assert(bM % w_m == 0, "bM %% w_m");
+        static_assert(bN % w_k == 0, "bN %% w_k");
+
+        constexpr int tile_dim_m = bM / w_m;
+        constexpr int tile_dim_k = bN / w_k;
+
+        constexpr int total_tiles = tile_dim_k * tile_dim_m;
+        constexpr int n_warp = block_y * block_z / 32;
+        constexpr int warp_cycle = (total_tiles + n_warp - 1) / n_warp;
+
+        float block_rescale_factor = 1.0f;
+        if constexpr (rescale) {
+          if constexpr (fixed) {
+            block_rescale_factor = scale_inv > 0 ? 65504.0f / (scale_inv * fixedMaxValue<T>::value) : 1.0f;
+          } else {
+            float thread_max = 0.0f;
+#pragma unroll
+            for (int c = 0; c < warp_cycle; c++) {
+              int logical_warp_index = c * n_warp + warp_id;
+              if (logical_warp_index < total_tiles) {
+#pragma unroll
+                for (int p = 0; p < pattern.get_number_phases(); p++) {
+                  int warp_m = (c * n_warp + warp_id) % tile_dim_m;
+                  int warp_k = (c * n_warp + warp_id) / tile_dim_m;
+
+                  int smem_m_offset = warp_m * w_m + pattern.get_m(lane_id, p);;
+                  int smem_k_offset = warp_k * w_k + pattern.get_n(lane_id, p);;
+
+                  int gmem_m_offset = smem_m_offset;
+                  int gmem_k_offset = smem_k_offset;
+
+                  float this_max = find_abs_max<x, fixed, dagger, tmp_ld>(load_t{}, smem_ptr, gmem_m_offset, gmem_k_offset, scale_inv);
+                  thread_max = fmaxf(this_max, thread_max);
+                }
+              }
+            }
+
+            // block all-reduce thread_max
+            using block_reduce_t = cub::BlockReduce<float, 1, cub::BLOCK_REDUCE_WARP_REDUCTIONS, block_y, block_z>;
+            __shared__ typename block_reduce_t::TempStorage temp_storage;
+            float block_max = block_reduce_t(temp_storage).Reduce(thread_max, cub::Max());
+
+            __shared__ float block_max_all;
+            if (threadIdx.x + blockDim.x * (threadIdx.y + blockDim.y * threadIdx.z) == 0) {
+              if (block_max > 0.0f) {
+                block_max_all = block_max;
+              } else {
+                block_max_all = 1.0f;
+              }
+            }
+            __syncthreads();
+
+            block_rescale_factor = 65504.0f / block_max_all; // 65504 = the maximum FP16 number
```

Comment:
I don't love `65504.0f` just being a "magic number" here, but I don't think there's a better place to put it---elsewhere in QUDA we use `std::numeric_limits`, but there isn't a `half_t` yet (it's not coming until C++23). Maybe we should create a custom `quda::numeric_limits` that calls `std::numeric_limits` where it's C++ legal and then has custom overloads for `half`? Still awkward because we use `short` as a magic way to refer to `half`, I guess... thoughts?

Reply:
Maybe it could just be something in `gmem_loader.cuh`?

Reply:
Does libcu++ have this already?  It often back ports things to earlier, and may already support `numeric_limits` for type `T = __half` or some such.

Reply:
I have added `numeric_limits<half>` in `gmem_loader.cuh`, for now, in https://github.com/lattice/quda/pull/1497/commits/6451e7fb56974b902baacdb64f547ef8eb5231c4.